eva/executor/create_index_executor.py

```diff
@@ -0,0 +1,213 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from pathlib import Path
+
+import faiss
+import pandas as pd
+
+from eva.catalog.catalog_manager import CatalogManager
+from eva.catalog.catalog_type import ColumnType, Dimension, NdArrayType, TableType
+from eva.catalog.index_type import IndexType
+from eva.configuration.constants import EVA_DEFAULT_DIR, INDEX_DIR
+from eva.executor.abstract_executor import AbstractExecutor
+from eva.models.storage.batch import Batch
+from eva.parser.create_statement import ColConstraintInfo, ColumnDefinition
+from eva.planner.create_index_plan import CreateIndexPlan
+from eva.sql_config import IDENTIFIER_COLUMN
+from eva.storage.storage_engine import StorageEngine
+from eva.utils.logging_manager import logger
+
+
+class CreateIndexExecutor(AbstractExecutor):
+    def __init__(self, node: CreateIndexPlan):
+        super().__init__(node)
+
+    def validate(self):
+        pass
+
+    def exec(self):
+        catalog_manager = CatalogManager()
+        if catalog_manager.get_index_by_name(self.node.name):
+            msg = f"Index {self.node.name} already exists."
+            logger.error(msg)
+            raise RuntimeError(msg)
+
+        try:
+            # Get feature tables.
+            table_ref = self.node.table_ref
+            df_metadata = catalog_manager.get_dataset_metadata(
+                table_ref.table.database_name, table_ref.table.table_name
+            )
+
+            # Add features to index.
+            # TODO: batch size is hardcoded for now.
+            index = None
+            input_dim, feat_size = -1, 0
+            logical_to_row_id = dict()
+            storage_engine = StorageEngine.factory(df_metadata)
+            for batch in storage_engine.read(df_metadata, 1):
+                # Feature column name.
+                feat_col_name = self.node.col_list[0].name
+
+                # Pandas wraps numpy array as an object inside a numpy
+                # array. Use zero index to get the actual numpy array.
+                feat = batch.column_as_numpy_array(feat_col_name)[0]
+                if index is None:
+                    input_dim = feat.shape[-1]
+                    index = self._create_index(self.node.index_type, input_dim)
+                index.add(feat)
+
+                # Secondary mapping.
+                row_id = batch.column_as_numpy_array(IDENTIFIER_COLUMN)[0]
+                logical_to_row_id[feat_size] = row_id
+
+                feat_size += 1
+
+            # Input dimension is inferred from the actual feature.
+            input_index_io = catalog_manager.index_io(
+                "input_feature",
+                ColumnType.NDARRAY,
+                NdArrayType.FLOAT32,
+                [Dimension.ANYDIM, input_dim],
+                True,
+            )
+
+            # Output dimension can be hardcoded.
+            id_index_io = catalog_manager.index_io(
```

Comment:
I think it is better to leave it as ndarray type here, because users can possibly search multiple input feature vectors together or top N matches (instead of top 1).

Reply:
Ah! This corresponds to the output of the index.

Reply:
We should document the input outputs somewhere. The best place would be the catalog file `index.`